### PR TITLE
Fix field-tag of MachinePool TaintIndentifier

### PR DESCRIPTION
**Pull request by suhanime**
This commit fix 2 typos:
- Field-tags for MachinePool TaintIdentifier had a colon before omitempty instead of a comma
- When making a slice for OwnedTaints as a part of update, set it to the length of unique taints instead of MachinePool spec.taints, as Machinepool could have duplicate taints which will not be migrated to OwnedTaints

xref: https://issues.redhat.com/browse/HIVE-2268

/assign @2uasimojo

### CODE REVIEW

**Comment by suhanime** [comment]
Verified on a live cluster.
I created a cluster, and MachinePool status had:
```
"ownedLabels": [
    "foo"
],
"ownedTaints": [
   {
       "effect:omitempty": "NoSchedule",
       "key:omitempty": "foo"
   }
],
```
After updating the hive to deploy current changes, MachinePool status updated to reflect
```
"ownedLabels": [
    "foo"
],
"ownedTaints": [
   {
      "effect": "NoSchedule",
      "key": "foo"
  }
],
```
Also checked if everything worked as expected when I updated | removed | re-added the labels and taints in spec. Logs had no errors.

Since this is a change in status field of MachinePool, and verified on a live cluster, it is safe to assume this should be able to handle existing clusters without breaking things

File: pkg/controller/machinepool/machinepool_controller.go [hunk 1]
```diff
@@ -960,8 +960,9 @@ func (r *ReconcileMachinePool) updatePoolStatusForMachineSets(
 	}
 
 	// ...and taints
-	pool.Status.OwnedTaints = make([]hivev1.TaintIdentifier, len(pool.Spec.Taints))
-	for i, taint := range *controllerutils.GetUniqueTaints(&pool.Spec.Taints) {
+	uniqueTaints := *controllerutils.GetUniqueTaints(&pool.Spec.Taints)
+	pool.Status.OwnedTaints = make([]hivev1.TaintIdentifier, len(uniqueTaints))
```
Comment (2uasimojo) [inline pkg/controller/machinepool/machinepool_controller.go:964]: What was the effect before this fix? Would OwnedTaints end up with trailing empty entries at the end?

```yaml
  ownedTaints:
  - key: foo
    effect: bar
  - key: ""
    effect: ""
```
?

Can we add unit test coverage for this case? I guess by adding a length check, and/or verifying all taintEffects are non-"empty"?

Comment (suhanime) [reply pkg/controller/machinepool/machinepool_controller.go:976]: Okay so you were totally correct, there were empty entries just like you mentioned. I added some unit tests to ensure we don't run into it again - I took the easy way out and split the setting of OwnedLabels and OwnedTaints in a separate function. I'm also explicit about length check to not run into nil vs empty comparisons.

**Review by 2uasimojo — COMMENTED** [review]
Code side is good.

/lgtm

Tests could be tidied up a bit, but doesn't need to block this -- cancel the hold if you prefer to do that later.

/hold

File: pkg/controller/machinepool/machinepool_controller_test.go [hunk 2]
```diff
@@ -1311,6 +1311,88 @@ func sortedTaints(taints []corev1.Taint) []corev1.Taint {
...
+				},
+			},
+		},
+		{
+			name: "Remove duplicates",
+			machinePool: func() *hivev1.MachinePool {
+				mp := testMachinePoolWithoutLabelsTaints()
+				mp.Spec.Labels = make(map[string]string)
+				mp.Spec.Labels["test-label"] = "test-value"
+				mp.Spec.Labels["test-label"] = "test-value"
```
[the remark this answers is not in the document]
Comment (suhanime) [reply pkg/controller/machinepool/machinepool_controller_test.go:1355]: Actually this isn't needed at all, dropping

File: pkg/controller/machinepool/machinepool_controller_test.go [hunk 3]
```diff
@@ -1311,6 +1311,88 @@ func sortedTaints(taints []corev1.Taint) []corev1.Taint {
...
+				mp.Spec.Taints = append(mp.Spec.Taints, corev1.Taint{
+					Key:    "test-taint",
+					Value:  "test-value",
+					Effect: "NoSchedule",
+				})
+				mp.Spec.Taints = append(mp.Spec.Taints, corev1.Taint{
+					Key:    "test-taint",
+					Value:  "test-value",
```
[the remark this answers is not in the document]
Comment (suhanime) [reply pkg/controller/machinepool/machinepool_controller_test.go:1363]: There is a similar test in Reconcile tests, but I don't mind proving it again. Coming right up

**Comment by 2uasimojo** [comment]
e2e-pool fail was an infra flake after the actual test succeeded. Merging.

**Pull request merged**